treewide: enable working, fix fixable, disable broken tests

oxij opened the pull request:
###### Motivation for this change

#33599 ("Run check phase by default"). This is yet another^4 (v2) followup to #39464 ("stdenv: implement most of #33599").

###### Things done

- Writing from a system with these changes applied on top of SLNOS (NotchB) and `doCheckByDefault` set, so I think it should work sane enough on top of Nixpkgs too.
- But I have not built everything this touches yet (in progress) and I still have some patches on the topic that might have influenced my builds, but they need some more work and I'm out of steam on this ATM.

###### Expectations

- When to merge: ASAP, but not earlier than in 24 hours.
- If I'm MIA and this starts conflicting with something just skip conflicting patches from this branch, I'll respin them later.

oxij committed:
perlPackages.JSON: fix tests

oxij committed:
jdupes: disable tests

oxij committed:
qt5.qtwebkit: disable tests

oxij committed:
qt5.qtwebkit.hyphen: fix tests

oxij committed:
pango: disable tests

oxij committed:
pmount: disable tests

oxij committed:
texinfo: use checkInputs

oxij committed:
libndctl: fix tests

oxij committed:
volume_key: disable tests

oxij committed:
radiotray-ng: use checkInputs, but disable tests, also cleanup

oxij committed:
feh: use checkInputs, fix tests

oxij committed:
rsstail: use checkInputs

oxij committed:
utox: use checkInputs, also cleanup

oxij committed:
fossil: cleanup tests

oxij committed:
mr: fix tests

oxij committed:
subversion: use checkInputs, make tests run, but them keep disabled

oxij committed:
treewide: random cleanups

oxij committed:
gnome3.gnome-keyring: use checkInputs

oxij committed:
cmdstan: use checkInputs

oxij committed:
glslang: disable tests

oxij committed:
vala: make tests run, but disable

oxij committed:
guile-cairo: use checkInputs

oxij committed:
apr: fix most tests

oxij committed:
clucene_core_2: disable tests

oxij committed:
fftw: use checkInputs, fix tests

oxij committed:
glog: make tests run, but disable

oxij committed:
gmime2: use checkInputs, fix tests

oxij committed:
gmime3: use checkInputs, fix tests

oxij committed:
gpgme: make tests run, but disable. Spooky!

oxij committed:
gst_all_1.gst-plugins-base: disable tests

oxij committed:
gst_all_1.gst-plugins-good: disable tests

oxij committed:
gd: disable tests

oxij committed:
gstreamer: make tests run, but disable, also cleanup

oxij committed:
harfbuzz: fix tests, also cleanup

oxij committed:
hunspell: fix tests

oxij committed:
jbig2dec: make tests run, but disable

oxij committed:
ldns: make tests run, but disable

oxij committed:
libappindicator: disable tests

oxij committed:
libdbusmenu: disable tests

oxij committed:
libdevil: fix tests

oxij committed:
libdrm: fix tests

oxij committed:
libfaketime: use checkInputs, fix tests

oxij committed:
libgsf: use checkInputs

oxij committed:
libminc: cleanup tests

oxij committed:
libosinfo: use checkInputs

oxij committed:
librdf: disable tests

oxij committed:
libsecret: make tests run, but disable, also cleanup

oxij committed:
libsoup: cleanup

oxij committed:
libtoxcore: cleanup tests

oxij committed:
liburcu: use checkInputs

oxij committed:
libzip: fix tests

oxij committed:
lightning: use checkInputs

oxij committed:
nlohmann_json: cleanup tests

oxij committed:
nss: disable tests

oxij committed:
openssl: fix tests, also cleanup

oxij committed:
pixman: libpng is an optional dependency, not a member of `checkInputs`

Since it was used before, let's keep using it.

oxij committed:
graphite2: make tests run, but disable

oxij committed:
sqlite: disable tests

oxij committed:
twolame: disable tests

oxij committed:
uthash: use checkInputs

oxij committed:
vigra: disable tests

oxij committed:
libpng: cleanup tests

oxij committed:
pythonPackages.dbus-python: use checkInputs

oxij committed:
pythonPackages.gensim: use checkInputs

oxij committed:
binutils: disable installcheck for different targetPlatforms

oxij committed:
elfutils: make tests run, but disable. A bit spooky!

oxij committed:
intel-gpu-tools: fix tests

oxij committed:
rman: make tests run, but disable

oxij committed:
strace: disable tests. Spooky!

oxij committed:
bison: use checkInputs

oxij committed:
flex, flex_2_6_1: fix tests

oxij committed:
re2c: fix tests

oxij committed:
nodejs: make tests run, but disable

oxij committed:
foo2zjs: make tests run, but disable

oxij committed:
cryptsetup: fix tests

oxij committed:
fuse: make tests run, but disable

oxij committed:
libaio: fix tests

oxij committed:
numactl: make tests run, but disable. Ugh!

oxij committed:
syslinux: make tests run, but disable

oxij committed:
utillinux: make tests run, but disable

oxij committed:
knot-resolver: use checkInputs

oxij committed:
lighttpd: make tests run, but disable

oxij committed:
myserver: use checkInputs

oxij committed:
redis: disable tests

oxij committed:
xcbutilxrm: use checkInputs

oxij committed:
zstd: make tests run, but disable

oxij committed:
e2fsprogs: make tests run, but disable

oxij committed:
pythonPackages.nixpart: use checkInputs

oxij committed:
sshfs-fuse: use checkInputs, fix tests

oxij committed:
memtest86plus: disable tests

oxij committed:
parted: use checkInputs

oxij committed:
recutils: use checkInputs, fix tests, also fix build

oxij committed:
system-config-printer: disable tests

oxij committed:
aria2: make tests run, but disable

oxij committed:
chrony: fix tests

oxij committed:
mailutils: use checkInputs, make tests run, but disable, also cleanup

oxij committed:
socat: make tests run, but disable

oxij committed:
swec: use checkInputs

oxij committed:
tcpdump: use checkInputs, fix tests

oxij committed:
nix-info: use checkInputs

oxij committed:
paperkey: fix tests

oxij committed:
pass: convert check to installcheck, fix tests

oxij committed:
gptfdisk: fix tests

oxij committed:
kakasi: make tests run, but disable

oxij committed:
wdiff: use checkInputs, fix tests

oxij committed:
xmlto: fix tests

oxij committed:
pythonPackages.cassandra-driver: use checkInputs

oxij committed:
netcdf: fix most tests

oxij committed:
nifticlib: disable tests

matthewbauer commented on pkgs/development/compilers/vala/default.nix:
```diff
@@ -26,6 +27,19 @@ let
+    #   sed -i -e 's@dbus/[^ ]*\.\(test\|vala\)@@g' tests/Makefile.am
+    # '';
+    #
+    # checkInputs = [ automake_1_15 ];
```
This looks like we should just leave off it isn't working.

matthewbauer commented on pkgs/development/libraries/libdevil/default.nix:
```diff
@@ -42,6 +42,13 @@ stdenv.mkDerivation rec {
+  postPatch = ''
+    for a in test/Makefile.in test/format_test/format_checks.sh.in ; do
+      substituteInPlace $a \
+        --replace /bin/bash ${bash}/bin/bash
```
I think it's considered more portable to do ```stdenv.shell``` instad of ```${bash}/bin/bash.

matthewbauer commented on pkgs/development/libraries/libpng/default.nix:
```diff
@@ -26,9 +25,7 @@ in stdenv.mkDerivation rec {
   propagatedBuildInputs = [ zlib ];
 
-  # it's hard to cross-run tests and some check programs didn't compile anyway
-  makeFlags = stdenv.lib.optional (!doCheck) "check_PROGRAMS=";
```
This might still be needed when stdenv.hostPlatform /= stdenv.buildPlatform;

jtojnar commented on pkgs/desktops/gnome-3/core/gnome-keyring/default.nix:
```diff
@@ -22,14 +22,11 @@ stdenv.mkDerivation rec {
   ];
 
-  # In 3.20.1, tests do not support Python 3
-  checkInputs = [ dbus python2 ];
```
Could you keep the comment?

matthewbauer commented on pkgs/tools/misc/parted/default.nix:
```diff
@@ -45,10 +44,12 @@ stdenv.mkDerivation rec {
+
+  # The `t0400-loop-clobber-infloop.sh' test wants `mkswap'.
+  preCheck = ''
+    export PATH=${utillinux}/sbin:$PATH
```
This should no longer be needed, sbin is a symlink to bin.

matthewbauer reviewed (commented):
Looks good aside from a few things.

oxij committed:
flex_2_5_35: make tests run, but disable

oxij committed:
gst-plugins-good: disable tests

oxij committed:
knot-dns: enable check, disable installcheck

oxij committed:
libseccomp: make tests run, but dependency cycle

oxij committed:
bash: make tests run, but dependency cycle

oxij committed:
pcre2: disable tests

oxij committed:
patchelf: disable tests

oxij committed:
automake111x: disable tests, also cleanup

oxij committed:
automake116x: disable tests, also cleanup

oxij committed:
cups: fix tests

oxij committed:
libcdio: fix tests

oxij committed:
gnome3.gcr: make tests run, but disable

oxij committed:
recode: disable tests

oxij committed:
glibmm: disable tests

oxij commented:
I rebased the fixes to comments and added all the required patches to build all the packages this touches with `doCheckByDefault` enabled (unless they conflicted, those I will respin later). I also built all of said packages.

So, in short, I think this should be ready to merge, but even if you see no issues, please wait till noon UTC tomorrow, just in case.

@matthewbauer

> This looks like we should just leave off it isn't working.

Yep, removed.

> I think it's considered more portable to do ```stdenv.shell``` instad of ```${bash}/bin/bash.

Done.

> This might still be needed when stdenv.hostPlatform /= stdenv.buildPlatform;

Nope, before this `doCheck` was set to `true` so that code was never ever run.

@jtojnar

> Could you keep the comment?

Done.

The original version is kept as `docheck/continues-in-a-big-way-v1` branch in my repo for `diff`ing.

matthewbauer reviewed (approved):
Looks good! Make sure you follow the Hydra staging jobset after this merges. Occasionally new tests will cause headaches on the Hydra infrastructure.

edolstra commented:
In "Motivation for this change", can you please describe the actual motivation rather than point to different PRs/issues? For example, this PR's motivation refers to #39464, which helpfully is entitled "stdenv: implement most of #33599" and whose motivation says "see #39463". That's not very helpful with a PR that causes an enormous amount of code churn (127 files changed).

oxij commented:
@edolstra >.< IMHO This PR's title clearly describes what this does and combined with "<package name>: (fix|disable) tests"-like commit messages I'm hard pressed to come up with something better.

> actual motivation

Sorry, but I can not be not snarky about this, either don't take this personally, or ignore the rest of the comment.

Eh? "I want to run tests provided by package authors, I want packages explicitly specify when some of their tests are broken, I don't want broken tests to prevent me from building the rest of my system until said tests are fixed." Duh? IMHO clicking the #33599 link would've saved you from writing that comment.

I copied issue titles into the OP (for people that can't click more than one layer deep, I guess) and edited #39464's OP to be a bit clearer (that one could've been better, true, but that striked out OP was written that way because #39464 is the central PR, but #39463 was PRed before it, and so it made sense to do it that way at the time).

oxij commented:
I looked over this with a fresh head and could find only several minor OCD things.

Old versions are kept as `docheck/continues-in-a-big-way-v2` (the version before today) and `docheck/continues-in-a-big-way-v2-ocd` (the version with ocd spelled out) for diffing. I squashed OCD patches into the current version.

It merges with staging, it evaluates, everything compiles. LGTM. It's perfect. Let's merge.

oxij commented:
ping @matthewbauer @Ericson2314 @7c6f434c

Could somebody please merge this? I think this can only get worse, not better, with coming conflicts.

7c6f434c commented:
Seems fine, and with all the semi-auto-updates we need the tests eventually enabled…
